### DISCUSSION TITLE
[General] signature-checker library not validating the signature

### DISCUSSION THREAD

**Discussion by lutr0** [zkSync-Community-Hub/zksync-developers#156] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I am trying to validate on chain using @matterlabs/signature-checker a message signed off-chain 

### What is the specific issue or error you're encountering?

I am able to successfully sign and verify a signature off-chain with `isMessageSignatureCorrect`, while on-chain `isValidSignatureNow` fails when passing the same hashed message and signature. 

### Can you share the error messages or logs you're receiving, if any?

```
ERC20Paymaster
Message hash:  0xc0f33fe245593955e70b011c20a2c606b7299b3b8aea1702f15181225f94b5da
Signed message hash:  0xaf083395c6ade025201eafd5967db4b03c6576fea54582d8e7caf8c4dfb8d44d4e600de9d360c3162c5e30513140b39e9b07f27aa3f62599edf78a4f988bd3621c
User address:  0x14764BD605657213CaDF0F153fd389Fc7496c2B5
ERC20 address:  0x04e1d9C8036B2D3250F68177333167A38236c2E1
Minimal allowance:  1
Max fee per gas:  [Function (anonymous)]
Gas limit:  6000000
Inner input:  0xaf083395c6ade025201eafd5967db4b03c6576fea54582d8e7caf8c4dfb8d44d4e600de9d360c3162c5e30513140b39e9b07f27aa3f62599edf78a4f988bd3621c
Paymaster params:  {
  paymaster: '0x3471092D4da0F8C6bD0597C50d07742aaB1b7256',
  paymasterInput: '0x949431dc00000000000000000000000004e1d9c8036b2d3250f68177333167a38236c2e1000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000041af083395c6ade025201eafd5967db4b03c6576fea54582d8e7caf8c4dfb8d44d4e600de9d360c3162c5e30513140b39e9b07f27aa3f62599edf78a4f988bd3621c00000000000000000000000000000000000000000000000000000000000000'
}
    1) Verify signature


  0 passing (1s)
  1 failing

  1) ERC20Paymaster
       Verify signature:
     Error: processing response error (body="{\"jsonrpc\":\"2.0\",\"error\":{\"code\":3,\"message\":\"Execution error: Failed to execute next transaction: Paymaster validation error: Invalid signature\",\"data\":\"0xeef172ea1435f2b21111953414267ed04ce77b61f7412521c799e517fc2983f4\"},\"id\":115}\n", error={"code":3,"data":"0xeef172ea1435f2b21111953414267ed04ce77b61f7412521c799e517fc2983f4"}, requestBody="{\"method\":\"eth_sendRawTransaction\",\"params\":[\"0x71f901c88080840ee6b280835b8d809404e1d9c8036b2d3250f68177333167a38236c2e180b84440c10f1900000000000000000000000014764bd605657213cadf0f153fd389fc7496c2b5000000000000000000000000000000000000000000000000000000000000000582010480808201049414764bd605657213cadf0f153fd389fc7496c2b582c350c0b84188257d118a52ae83fd20b6e4bd7743c33397f4aa5c2be28ca6ce5c961d165b95557062614db9557c6def1394dcff8c668531e4e692266b669ad8bf42c59f5f861bf8fb943471092d4da0f8c6bd0597c50d07742aab1b7256b8e4949431dc00000000000000000000000004e1d9c8036b2d3250f68177333167a38236c2e1000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000041af083395c6ade025201eafd5967db4b03c6576fea54582d8e7caf8c4dfb8d44d4e600de9d360c3162c5e30513140b39e9b07f27aa3f62599edf78a4f988bd3621c00000000000000000000000000000000000000000000000000000000000000\"],\"id\":115,\"jsonrpc\":\"2.0\"}", requestMethod="POST", url="http://localhost:8011", code=SERVER_ERROR, version=web/5.7.1)
      at Logger.makeError (node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
      at Logger.throwError (node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
      at /home/joe/Dev/zksync-signature-checker/node_modules/@ethersproject/web/src.ts/index.ts:341:28
      at step (node_modules/@ethersproject/web/lib/index.js:33:23)
      at Object.next (node_modules/@ethersproject/web/lib/index.js:14:53)
      at fulfilled (node_modules/@ethersproject/web/lib/index.js:5:58)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)



error Command failed with exit code 1.
```

### Have you made any recent changes to the contract before encountering this issue?

N/A

### Are there any external libraries or contracts that your contract interacts with?

N/A

### Can you provide the relevant portions of your contract code where the issue is occurring?

I recreated a repo with the issue, please see [here](https://github.com/lutr0/zksync-signature-checker/blob/7a1bc92370a484814c1b59e45d7dc3d2985541f8/contracts/Paymaster.sol#L172)

### Have you tried to isolate the problem, and if so, what were the results?

yes, please see the repo. I am checking the signature succesfully off-chain and I am validating that I am passing the same parameters on-chain

### What steps have you already taken to try to resolve the issue?

N/A

### Repo Link (Optional)

https://github.com/lutr0/zksync-signature-checker

**Comment by dutterbutter** [comment]
Hey @lutr0 👋 

We are about to deprecate `@matterlabs/signature-checker/contracts/SignatureChecker.sol` in favour of open zeppelins signature-checker [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/5ed5a86d1d22f387ce69ab4e0ace405de8bc888d/contracts/utils/cryptography/SignatureChecker.sol#L22). Examples in our docs will be updated today to reflect this change.

**Reply by lutr0** [reply]
Hey @dutterbutter,
many thanks for your help! 
I added a [commit](https://github.com/lutr0/zksync-signature-checker/commit/ed24ca8145fb81d5b419121d4b6fc178b23c37a6) to run the code with the OZ library, and I am getting the same error.

As I said it's very strange because the same parameters work off-chain

![image](https://github.com/zkSync-Community-Hub/zkync-developers/assets/76906369/647a6379-5688-4043-9cc8-1c74b4b58d8d)

**Reply by dutterbutter** [reply]
Thanks, I can investigate further here.

**Reply by dutterbutter** [reply, marked as answer]
I've reviewed the repo and got the test to pass. My changes can be found on this fork [here](https://github.com/dutterbutter/zksync-signature-checker/commit/87d7a36aade9a89ee55b93588e94aa530f312ae9), but can be summarized as:

- When calling `verifier.signMessage`, the `messageHash` is now passed through `ethers.utils.arrayify` to ensure it's in the correct format for signing.
- Similarly, when setting the `innerInput` field of `paymasterParams`, `SignedMessageHash` is also passed through `ethers.utils.arrayify` to ensure correct formatting.
- Updated `isValidSignature` to make use of `ECDSA.toEthSignedMessageHash` and `ECDSA.tryRecover` methods. 

That should resolve the invalid signature issue.

**Reply by lutr0** [reply]
Thank you so much, this indeed fixed the issue!

**Comment by biotremenag** [comment]
It seems like there might be an issue with the signature verification process either on-chain or with the integration of the @matterlabs/signature-checker library. Here are a few steps you can take to troubleshoot and resolve the issue:

Double-check Parameters: Ensure that the parameters being passed for signature verification on-chain exactly match those used off-chain, including the message hash and the signature itself.

Library Compatibility: Confirm that the @matterlabs/signature-checker library is compatible with the version of zkSync you are using. It's possible that there could be compatibility issues leading to validation failures.

Debugging: Implement additional logging or debugging mechanisms in your code to track the flow of data and identify any discrepancies between off-chain and on-chain processes.

Consult Documentation: Review the documentation for @matterlabs/signature-checker and zkSync to ensure that you are following the correct procedures for signature verification both off-chain and on-chain.

Community Support: Reach out to the zkSync community or the developers of @matterlabs/signature-checker for assistance. They may be able to provide insights or guidance on resolving the issue.

By following these steps and carefully examining the code and processes involved in signature verification, you should be able to identify and address the underlying cause of the validation failure.

**Comment by davenportbri** [comment]
The issue you're encountering might be due to differences in the message hashing or signature verification process between off-chain and on-chain environments. Double-check the hashing algorithm and the signature verification method used in both cases to ensure consistency. Additionally, ensure that all parameters passed to the on-chain function match exactly what's expected, including any encoding or formatting details.

**Comment by loal0cksibus** [comment]
It seems like the error might be due to discrepancies between the off-chain signature verification and the on-chain validation process. Ensure that the hashing algorithm, message, and signature parameters are consistent across both processes. Additionally, verify that the contract's environment and dependencies are properly configured for signature verification.
